library/src/main/java/com/datatorrent/lib/join/AbstractInnerJoinOperator.java

```diff
@@ -0,0 +1,331 @@
+ * applies the join operation based on constraint and emit the joined value.
+ * Concrete classes should provide implementation to extractKey, extractTime, mergeTuples methods.
+ *
+ * <b>Properties:</b><br>
+ * <b>includeFieldStr</b>: List of comma separated fields to be added to the output tuple.
+ *                         Ex: Field1,Field2;Field3,Field4<br>
+ * <b>keyFields</b>: List of comma separated key field for both the streams. Ex: Field1,Field2<br>
+ * <b>timeFields</b>: List of comma separated time field for both the streams. Ex: Field1,Field2<br>
+ * <b>expiryTime</b>: Expiry time for stored tuples<br>
+ * <b>isStream1KeyPrimary</b>: : Specifies whether the stream1 key is primary or not<br>
+ * <b>isStream2KeyPrimary</b>: : Specifies whether the stream2 key is primary or not<br>
+ *
+ * <b> Example: </b> <br>
+ *  Left input port receives customer details and right input port receives Order details.
+ *  Schema for the Customer be in the form of {ID, Name, CTime}
+ *  Schema for the Order be in the form of {OID, CID, OTime}
+ *  Now, Join the tuples of Customer and Order streams where Customer.ID = Order.CID and the constraint is
+ *  matched tuples must have timestamp within 5 minutes.
+ *  Here, key Fields = ID, CID and Time Fields = CTime, OTime, expiryTime = 5 minutes </b> <br>
+ *
+ *  @displayName Abstract Inner Join Operator
+ *  @tags join
+ */
+public abstract class AbstractInnerJoinOperator<K,T> extends BaseOperator
+{
+  protected transient String[][] includeFields;
+  protected transient List<String> keyFields;
+  protected transient List<String> timeFields;
+  @AutoMetric
+  private long tuplesJoinedPerSec;
+  private double windowTimeSec;
+  private int tuplesCount;
+  @NotNull
+  private String keyFieldsStr;
+  @NotNull
+  private String includeFieldStr;
+  private String timeFieldsStr;
+  private Long stream1ExpiryTime;
+  private Long stream2ExpiryTime;
+  private boolean isStream1KeyPrimary = true;
+  private boolean isStream2KeyPrimary = true;
+  protected SpillableComplexComponent component;
+  protected Spillable.SpillableByteArrayListMultimap<K,T> stream1Data;
+  protected Spillable.SpillableByteArrayListMultimap<K,T> stream2Data;
+
+  /**
+   * Process the tuple which are received from input ports with the following steps:
+   * 1) Extract key from the given tuple
+   * 2) Insert <key,tuple> into the store where store is the stream1Data if the tuple
+   * receives from stream1 or viceversa.
+   * 3) Get the values of the key if found it in opposite store
+   * 4) Merge the given tuple and values found from step (3)
+   * @param tuple given tuple
+   * @param isStream1Data Specifies whether the given tuple belongs to stream1 or not.
+   */
+  protected void processTuple(T tuple, boolean isStream1Data)
+  {
+    Spillable.SpillableByteArrayListMultimap<K,T> store = isStream1Data ? stream1Data : stream2Data;
+    K key = extractKey(tuple,isStream1Data);
+    if (!store.put(key, tuple)) {
```

Comment:
If the tuple is already expired tuple then in this case will return false. put() will allow the duplicates.